java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -59,66 +59,78 @@ private class FromMkdirsFileFlowState extends DataFlow::FlowState {
  *
- * 'Safe' means that the directory is guarnteed to have been created, and the directory did not already exist.
+ * For example, gets the test expression in  `if(!f.mkdir()) { throw ... }` but not `if(!f.mkdir() && !f.exists()) { throw ... }`,
+ * since the latter accepts the case where `f` already exists.
  */
-private Guard safeDirectoryCreationGuard(ThrowStmt s) {
-  result = directoryCreationGuard(s) and
-  // This guard is not 'safe' if a `directoryExistsGuard` is also guarding this throw statement
-  not exists(directoryExistsGuard(s))
+private predicate throwsIfMkdirFailedExcludingExistenceChecks(Guard test) {
+  exists(ThrowStmt s |
+    throwsIfMkdirFailed(test, s) and
+    // The same throw statement must not result from a failing `f.exists()` or similar call.
+    not throwsIfDirectoryDoesNotExist(s)
+  )
 }
 
 /**
- * An expression that will create a directory without throwing an exception if a file/directory already exists.
+ * Holds if `creationCall` is of the form `fileInstanceExpr.mkdir()`,
+ * and it is not clear that failure to create a fresh directory will result in throwing an exception.
+ *
+ * Note `fileInstanceExpr` may be passed to the actual `mkdir` via a wrapper function: in this case
+ * `fileInstanceExpr` is the argument to the outer wrapper function, but `creationCall` is still the
+ * inner `mkdir` call.
+ *
+ * TODO: that may warrant changing.
  */
 private predicate isNonThrowingDirectoryCreationExpression(
   Expr fileInstanceExpr, MethodAccess creationCall
 ) {
   creationCall.getMethod() instanceof MethodFileCreatesDirs and
   creationCall.getQualifier() = fileInstanceExpr and
   (
-    // Check for 'throwing creations' and ensure that this call is not used in a throwing case.
-    // If the creation call is used in a guard:
-    creationCall.(Guard).directlyControls(_, _)
-    implies
-    // Then it must not be a 'safe' creation guard. Thus `creationCall` is not a 'throwing creation'.
-    not creationCall = safeDirectoryCreationGuard(_) // Ensure that the guard is insufficient.
+    // The result of `mkdir` is not directly tested at all...
+    not creationCall.(Guard).directlyControls(_, _)
+    or
+    // ... or it is tested, but that test doesn't clearly have the form `if(!f.mkdir()) throw ...`.
+    not throwsIfMkdirFailedExcludingExistenceChecks(creationCall)
   )
   or
   // Recursively look for methods that encapsulate the above.
   // Thus, the use of 'helper directory creation methods' are still considered
   // when assessing if an `unsafeUse` is present or not.
-  exists(Method m, int argumentIndex, Expr arg |
-    arg.(VarAccess).getVariable() = m.getParameter(argumentIndex) and
-    isNonThrowingDirectoryCreationExpression(any(Expr e2 | DataFlow::localExprFlow(arg, e2)),
-      creationCall)
+  exists(Method m, int argumentIndex, Expr fileInstanceParam |
+    DataFlow::localExprFlow(m.getParameter(pragma[only_bind_into](argumentIndex)).getAnAccess(),
```

Comment:
`pragma[only_bind_into](argumentIndex)` What does this do, and why use it in this case?

Reply:
https://codeql.github.com/docs/ql-language-reference/annotations/#pragma-only-bind-into

Basically I am trying to discourage a join on just `argumentIndex`, because this is an integer column with small cardinality which invites a very high cost join.

To give a simplified example, consider argument <-> parameter matching:

```
predicate argParamPair(Parameter p, Expr arg) {
  exists(MethodAccess ma, Method m, int i | m = ma.getMethod() and arg = ma.getArgument(i) and p = m.getParameter(i))
}
```

A bad solution ordering might look like:

```
Scan `MethodAccess.getArgument`, result `[methodAccess, i, argExpr]`
Join against `Method.getParameter` on matching `i`, result `[method, methodAccess, i, argExpr, parameter]`
Join against `MethodAccess.getMethod` on matching `method, methodAccess`, result `[argExpr, parameter]`
```

This is a bad choice because there are few values of `i`: in the worst possible case all our methods and therefore all  our method accesses have a single parameter/argument and "join on matching `i`" is a no-op, equivalent to a cartesian product of `MethodAccess.getArgument` and `Method.getParmeter`.

Generally CodeQL will not try to use a column as a join key until it has been bound (constrained) -- for primitive types this avoids a potentially infinite set of values, and for others it is simply good practice to ensure we don't join on a finite but large column and produce a very large intermediate result. By wrapping uses of `i` in `pragma[only_bind_into]` I tell CodeQL that evaluating `ma.getArgument(i)` shouldn't be regarded as usefully constraining `i`.

Precisely what it will do with that information I don't know exactly, but I'm hoping for an ordering like

```
Scan `MethodAccess.getArgument`, result `[methodAccess, i, argExpr]`
Join against `MethodAccess.getMethod` on matching `methodAccess`, result `[method, i, argExpr]`
Join against `Method.getParameter` on matching `method, i`, result `[argExpr, parameter]`
```